Panic in

# Required information

Ubuntu 16.04, LXD 2.0.2, LXC 2.0.2

# Issue description

The command `lxc remote add tmp tmp` will cause a panic if `tmp` is a valid path in the current working directory.  There are three distinct bugs related to the panic:

- Bad string comparison using unchecked slices (this is the immediate cause of the panic).
- Not checking if the path is a directory when determining scheme.  LXC requires the full path to the UNIX socket when communicating with a server, thus successful communication with a directory is impossible and the path match should be ignored, i.e. https:// scheme should be assumed.
- Handling of relative paths is inconsistent.  If relative paths are allowed, then client.go has a bug where it inadvertently converts relative paths to absolute paths.  If relative paths are not allowed, then the code should be hardened and explicit checks added to addServer and to client.go.

```
panic: runtime error: slice bounds out of range

goroutine 1 [running]:
panic(0x8c9080, 0xc82000e100)
	/usr/lib/go-1.6/src/runtime/panic.go:464 +0x3e6
main.(*remoteCmd).addServer(0xc820017900, 0xc8200e0330, 0x7fffed1330b9, 0x3, 0x7fffed1330bd, 0x3, 0x0, 0x0, 0x0, 0x44f100, ...)
	/home/sean/go/src/github.com/lxc/lxd/lxc/remote.go:151 +0x1cc6
main.(*remoteCmd).run(0xc820017900, 0xc8200e0330, 0xc820017ec0, 0x3, 0x4, 0x0, 0x0)
	/home/sean/go/src/github.com/lxc/lxd/lxc/remote.go:312 +0x576
main.run(0x0, 0x0)
	/home/sean/go/src/github.com/lxc/lxd/lxc/main.go:137 +0x1098
main.main()
	/home/sean/go/src/github.com/lxc/lxd/lxc/main.go:21 +0x27
```

# Steps to reproduce
`mkdir tmp`
`lxc remote add tmp tmp`

--- a/client.go
+++ b/client.go
@@ -255,11 +255,7 @@ func connectViaUnix(c *Client, remote *RemoteConfig) error {
 		//   unix:///path/to/socket
 		//   unix:/path/to/socket
 		//   unix:path/to/socket
-		path := strings.TrimPrefix(remote.Addr, "unix:")
-		if strings.HasPrefix(path, "///") {
-			// translate unix:///path/to, to just "/path/to"
-			path = path[2:]
-		}
+		path := strings.TrimPrefix(strings.TrimPrefix(remote.Addr, "unix:"), "//")
 		raddr, err := net.ResolveUnixAddr("unix", path)
 		if err != nil {
 			return nil, err

--- a/client_test.go
+++ b/client_test.go
@@ -23,7 +23,7 @@ func TestLocalLXDError(t *testing.T) {
 		Name:   "test",
 		Config: DefaultConfig,
 		Remote: &RemoteConfig{
-			Addr:   fmt.Sprintf("unix:/%s", f.Name()),
+			Addr:   fmt.Sprintf("unix:%s", f.Name()),
 			Static: true,
 			Public: false,
 		},

--- a/lxc/remote.go
+++ b/lxc/remote.go
@@ -126,7 +126,7 @@ func (c *remoteCmd) addServer(config *lxd.Config, server string, addr string, ac
 	} else if addr[0] == '/' {
 		rScheme = "unix"
 	} else {
-		if !shared.PathExists(addr) {
+		if !shared.IsUnixSocket(addr) {
 			rScheme = "https"
 		} else {
 			rScheme = "unix"
@@ -148,17 +148,7 @@ func (c *remoteCmd) addServer(config *lxd.Config, server string, addr string, ac
 	}
 
 	if rScheme == "unix" {
-		if addr[0:5] == "unix:" {
-			if addr[0:7] == "unix://" {
-				if len(addr) > 8 {
-					rHost = addr[8:]
-				} else {
-					rHost = ""
-				}
-			} else {
-				rHost = addr[6:]
-			}
-		}
+		rHost = strings.TrimPrefix(strings.TrimPrefix(addr, "unix:"), "//")
 		rPort = ""
 	}
 
@@ -181,7 +171,7 @@ func (c *remoteCmd) addServer(config *lxd.Config, server string, addr string, ac
 		return err
 	}
 
-	if len(addr) > 5 && addr[0:5] == "unix:" {
+	if strings.HasPrefix(addr, "unix:") {
 		// NewClient succeeded so there was a lxd there (we fingered
 		// it) so just accept it
 		return nil

--- a/shared/util.go
+++ b/shared/util.go
@@ -70,6 +70,16 @@ func IsDir(name string) bool {
 	return stat.IsDir()
 }
 
+// IsUnixSocket returns true if the given path is either a Unix socket
+// or a symbolic link pointing at a Unix socket.
+func IsUnixSocket(path string) bool {
+	stat, err := os.Stat(path)
+	if err != nil {
+		return false
+	}
+	return (stat.Mode() & os.ModeSocket) == os.ModeSocket
+}
+
 // VarPath returns the provided path elements joined by a slash and
 // appended to the end of $LXD_DIR, which defaults to /var/lib/lxd.
 func VarPath(path ...string) string {

